File: src-tauri/src/domain/error_masking.rs
const REDACTED: &str = "[REDACTED]";
const SENSITIVE_KEYS: [&str; 3] = ["password", "user id", "token"];

pub fn mask_sensitive_text(text: &str) -> String {
    mask_sensitive_text_with_values(text, &[])
}

pub(crate) fn mask_sensitive_text_with_values(text: &str, values: &[String]) -> String {
    let mut masked = mask_named_values(text);

    for value in values.iter().filter(|value| !value.is_empty()) {
        masked = replace_case_insensitive(&masked, value, REDACTED);
    }

    masked
}

fn mask_named_values(text: &str) -> String {
    let lower = text.to_ascii_lowercase();
    let mut result = String::with_capacity(text.len());
    let mut cursor = 0;

    while cursor < text.len() {
        let Some((key_start, key)) = next_sensitive_key(&lower, cursor) else {
            result.push_str(&text[cursor..]);
            break;
        };

        result.push_str(&text[cursor..key_start]);
        let key_end = key_start + key.len();
        let mut separator_end = key_end;
        while text
            .as_bytes()
            .get(separator_end)
            .is_some_and(u8::is_ascii_whitespace)
        {
            separator_end += 1;
        }

        if !matches!(text.as_bytes().get(separator_end), Some(b'=') | Some(b':')) {
            result.push_str(&text[key_start..key_end]);
            cursor = key_end;
            continue;
        }

        separator_end += 1;
        while text
            .as_bytes()
            .get(separator_end)
            .is_some_and(u8::is_ascii_whitespace)
        {
            separator_end += 1;
        }

        let value_end = end_of_sensitive_value(text, separator_end);
        result.push_str(&text[key_start..separator_end]);
        result.push_str(REDACTED);
        cursor = value_end;
    }

    result
}

fn next_sensitive_key<'a>(lower: &str, cursor: usize) -> Option<(usize, &'a str)> {
    SENSITIVE_KEYS
        .iter()
        .filter_map(|key| {
            lower[cursor..]
                .find(key)
                .map(|relative_start| (cursor + relative_start, *key))
        })
        .filter(|(start, _)| {
            *start == 0
                || !lower.as_bytes()[start - 1].is_ascii_alphanumeric()
                    && lower.as_bytes()[start - 1] != b'_'
        })
        .min_by_key(|(start, _)| *start)
}

fn end_of_sensitive_value(text: &str, start: usize) -> usize {
    match text.as_bytes().get(start) {
        Some(b'\'') | Some(b'\"') => {
            let quote = text.as_bytes()[start];
            text[start + 1..]
                .find(quote as char)
                .map_or(text.len(), |offset| start + offset + 2)
        }
        _ => text[start..]
            .find(|character: char| {
                character.is_whitespace() || matches!(character, ',' | ';' | '&' | ')')
            })
            .map_or(text.len(), |offset| start + offset),
    }
}

fn replace_case_insensitive(text: &str, value: &str, replacement: &str) -> String {
    let mut result = String::with_capacity(text.len());
    let lower = text.to_ascii_lowercase();
    let value_lower = value.to_ascii_lowercase();
    let mut cursor = 0;

    while let Some(relative_start) = lower[cursor..].find(&value_lower) {
        let start = cursor + relative_start;
        result.push_str(&text[cursor..start]);
        result.push_str(replacement);
        cursor = start + value.len();
    }

    result.push_str(&text[cursor..]);
    result
}

File: electron/domain/errorMasking.ts
const REDACTED = "[REDACTED]";
const SENSITIVE_KEYS = ["password", "user id", "token"] as const;

export function maskSensitiveText(text: string, secrets: readonly string[] = []): string {
  let masked = maskNamedValues(text);
  const longestFirst = secrets
    .filter((secret) => secret.length > 0)
    .slice()
    .sort((left, right) => right.length - left.length);

  for (const secret of longestFirst) {
    masked = replaceCaseInsensitive(masked, secret, REDACTED);
  }
  return masked;
}

function maskNamedValues(text: string): string {
  const lower = asciiLower(text);
  let result = "";
  let cursor = 0;

  while (cursor < text.length) {
    const sensitive = nextSensitiveKey(lower, cursor);
    if (sensitive === undefined) {
      result += text.slice(cursor);
      break;
    }

    const [keyStart, key] = sensitive;
    result += text.slice(cursor, keyStart);
    const keyEnd = keyStart + key.length;
    let separatorEnd = keyEnd;
    while (separatorEnd < text.length && isAsciiWhitespace(text[separatorEnd])) {
      separatorEnd += 1;
    }

    if (text[separatorEnd] !== "=" && text[separatorEnd] !== ":") {
      result += text.slice(keyStart, keyEnd);
      cursor = keyEnd;
      continue;
    }

    separatorEnd += 1;
    while (separatorEnd < text.length && isAsciiWhitespace(text[separatorEnd])) {
      separatorEnd += 1;
    }

    const valueEnd = endOfSensitiveValue(text, separatorEnd);
    result += text.slice(keyStart, separatorEnd);
    result += REDACTED;
    cursor = valueEnd;
  }

  return result;
}

function nextSensitiveKey(lower: string, cursor: number): [number, string] | undefined {
  let next: [number, string] | undefined;

  for (const key of SENSITIVE_KEYS) {
    let start = lower.indexOf(key, cursor);
    while (start >= 0 && !isSensitiveKeyBoundary(lower, start)) {
      start = lower.indexOf(key, start + 1);
    }
    if (start >= 0 && (next === undefined || start < next[0])) {
      next = [start, key];
    }
  }

  return next;
}

function isSensitiveKeyBoundary(lower: string, start: number): boolean {
  if (start === 0) {
    return true;
  }
  const previous = lower[start - 1];
  return !isAsciiAlphanumeric(previous) && previous !== "_";
}

function endOfSensitiveValue(text: string, start: number): number {
  const first = text[start];
  if (first === "'" || first === '"') {
    const closing = text.indexOf(first, start + 1);
    return closing < 0 ? text.length : closing + 1;
  }

  for (let index = start; index < text.length; index += 1) {
    const character = text[index];
    if (isWhitespace(character) || character === "," || character === ";"
      || character === "&" || character === ")") {
      return index;
    }
  }
  return text.length;
}

function replaceCaseInsensitive(text: string, value: string, replacement: string): string {
  const lower = asciiLower(text);
  const valueLower = asciiLower(value);
  let result = "";
  let cursor = 0;
  let start = lower.indexOf(valueLower, cursor);

  while (start >= 0) {
    result += text.slice(cursor, start);
    result += replacement;
    cursor = start + value.length;
    start = lower.indexOf(valueLower, cursor);
  }

  return result + text.slice(cursor);
}

function asciiLower(value: string): string {
  return value.replace(/[A-Z]/g, (character) => character.toLowerCase());
}

function isAsciiAlphanumeric(character: string): boolean {
  return /^[A-Za-z0-9]$/.test(character);
}

function isAsciiWhitespace(character: string): boolean {
  return character === " " || character === "\t" || character === "\n"
    || character === "\r" || character === "\f" || character === "\v";
}

function isWhitespace(character: string): boolean {
  return /\s/u.test(character);
}
